libcontainer/userns/userns_linux.go

```diff
@@ -15,22 +16,33 @@ var (
 // Originally copied from github.com/lxc/lxd/shared/util.go
 func runningInUserNS() bool {
 	nsOnce.Do(func() {
-		uidmap, err := user.CurrentProcessUIDMap()
+		file, err := os.Open("/proc/self/uid_map")
 		if err != nil {
 			// This kernel-provided file only exists if user namespaces are supported
 			return
 		}
-		inUserNS = uidMapInUserNS(uidmap)
+		defer file.Close()
+
+		buf := bufio.NewReader(file)
+		l, _, err := buf.ReadLine()
+		if err != nil {
+			return
+		}
+
+		inUserNS = uidMapInUserNS(string(l))
 	})
 	return inUserNS
 }
 
-func uidMapInUserNS(uidmap []user.IDMap) bool {
+func uidMapInUserNS(uidMap string) bool {
+	var a, b, c int64
+	_, err := fmt.Sscanf(uidMap, "%d %d %d", &a, &b, &c)
+
 	/*
 	 * We assume we are in the initial user namespace if we have a full
 	 * range - 4294967295 uids starting at uid 0.
 	 */
-	if len(uidmap) == 1 && uidmap[0].ID == 0 && uidmap[0].ParentID == 0 && uidmap[0].Count == 4294967295 {
+	if err == nil && a == 0 && b == 0 && c == 4294967295 {
 		return false
 	}
 	return true
```

Comment:
Made this change per suggestion in https://github.com/opencontainers/runc/pull/2868#discussion_r617925082

Looking at the logic now, I'm wondering if `return true` is "correct" if we have an error; is it ok to assume that we are in userns if we receive an error, or should we assume the reverse?

I guess if this file is not parsable, it means things are non-functional anyway, so perhaps not an issue (but looking from the other perspective; if the file-mapping is not-parsable, we likely won't be able to be using user-namespaces, in which case `false` would make more sense)

@AkihiroSuda @kolyshkin wdyt?

Reply:
Alternative would be:

```go
func uidMapInUserNS(uidMap string) bool {
	var a, b, c int64
	_, err := fmt.Sscanf(uidMap, "%d %d %d", &a, &b, &c)
	if err != nil {
		return false
	}

	/*
	 * We assume we are in the initial user namespace if we have a full
	 * range - 4294967295 uids starting at uid 0.
	 */
	if a == 0 && b == 0 && c == 4294967295 {
		return false
	}
	return true
}
```

Reply:
One other way would be to `panic()` if we can't parse this. That is too radical I think.

As to whether return `true` or `false` in case we're not sure, I'd say emit a warning and return `false` as e.g. cgroups/fs drivers use `true` to skip devices setting, and we don't want that.

So I like your last version, which can be further simplified by doing
```go
initns := a == 0 && b == 0 && c == 4294967295
return !initns
```

I'd also replace the comment ("We assume we are ...") with a reference to `user_namespaces(7)`:

```
// As per user_namespaces(7), /proc/self/uid_map of
// the initial user namespace shows 0 0 4294967295.
```

Reply:
Yes, I think in lxc, they don't check `uidMap == ""`, so they return `true` when got an `EOF` error.
But we checked this, so `return false ` is correct, I think it will never happen.